src/OrleansRuntime/Versions/CachedVersionSelectorManager.cs

```diff
@@ -0,0 +1,54 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using Orleans.Runtime.Versions.Compatibility;
+using Orleans.Runtime.Versions.Selector;
+
+namespace Orleans.Runtime.Versions
+{
+    internal class CachedVersionSelectorManager
+    {
+        private readonly GrainTypeManager grainTypeManager;
+        private readonly Func<Tuple<int, int, ushort>, IReadOnlyList<SiloAddress>> getSilosFunc;
+        private ConcurrentDictionary<Tuple<int,int,ushort>, IReadOnlyList<SiloAddress>> suitableSilosCache;
+
+        public VersionSelectorManager VersionSelectorManager { get; }
+
+        public CompatibilityDirectorManager CompatibilityDirectorManager { get; }
+
+        public CachedVersionSelectorManager(GrainTypeManager grainTypeManager, VersionSelectorManager versionSelectorManager, CompatibilityDirectorManager compatibilityDirectorManager)
+        {
+            this.grainTypeManager = grainTypeManager;
+            this.VersionSelectorManager = versionSelectorManager;
+            this.CompatibilityDirectorManager = compatibilityDirectorManager;
+            this.getSilosFunc = GetSuitableSilosImpl;
+        }
+
+        public IReadOnlyList<SiloAddress> GetSuitableSilos(int typeCode, int ifaceId, ushort requestedVersion)
+        {
+            var key = Tuple.Create(typeCode, ifaceId, requestedVersion);
+            return suitableSilosCache.GetOrAdd(key, getSilosFunc);
```

Comment:
It doesn't have a GetOrAdd method yet, but you could use CachedReadConcurrentDictionary instead here and have cheaper access times while retaining thread safety (and no need to reassign the dictionary when resetting the cache.

`ResetCache` below could use `.Clear()`